### DISCUSSION TITLE
Improved context info (caller identification), perhaps using Microsoft.CodeAnalysis

### DISCUSSION THREAD

**Discussion by gliljas** [fluentassertions/fluentassertions#1583] [Ideas]
I've experimented a bit with using Microsoft.CodeAnalysis in order to improve the caller identification. So far it looks quite promising, and while the same thing could certainly be done using only string parsing, traversing SyntaxNodes is quite convenient and also permits line breaks etc. while scanning for the caller. Maybe it's overkill, but I think it would at least be nice to be able to swap out the CallerIdentifier or just ExtractVariableNameFrom. 

Another thing that would be nice is to be able to augment the context for things like WhichValue. Currently, a failed assertion like:

`dictionary.Should().ContainKey("Key").WhichValue.Should().Be(4);`

will render a message like:

`Expected dictionary to be 4`

but we really didn't expect **dictionary** to be 4, but rather **dictionary["Key"]**.

**Comment by dennisdoomen** [comment]
Well, in this case, the message will have been `Expected dictionary to contain key 4`. But in any case, we're open to improvements. Maybe a spike PR?

**Reply by gliljas** [reply]
I'll create a PR, or two.

The message if the value is wrong (not the key) will be "Expected dictionary to be 4"

**Reply by dennisdoomen** [reply]
Ah, yes. That's actually a bug or design flaw. The `WhichValue` should change the subject identifier to something that refers to the pair with key `key`.
